### PR TITLE
feat(sdk): `EventCache` fully uses `RoomEvents`/`LinkedChunk`

### PR DESCRIPTION
**Pull request by Hywan**
`EventCache` uses a custom `EventCacheStore` trait to store all events for all rooms.

The idea of this PR is to use the newly introduced `LinkedChunk` in https://github.com/matrix-org/matrix-rust-sdk/pull/3166. Instead of having one store for all events for all rooms, now there is one `LinkedChunk`/store per room.

Of course, @bnjbvr and I are discussing a lot, which explains why the actual `EventCacheStore` was already adopting some patterns from `LinkedChunk`. The migration was easier.

Several PR were necessary to make this work possible:

* [x] Build on top of https://github.com/matrix-org/matrix-rust-sdk/pull/3227
* [x] Extracted commits about `LinkedChunk` in https://github.com/matrix-org/matrix-rust-sdk/pull/3236
* [x] Other extracted commits about `LinkedChunk` in https://github.com/matrix-org/matrix-rust-sdk/pull/3251

Briefly, this PR removes the `EventCacheStore` and the `MemoryStore`, and replace them by `LinkedChunk`.

---

* Address https://github.com/matrix-org/matrix-rust-sdk/issues/3058

### CODE REVIEW

**Comment by Hywan** [comment]
Marking the PR as _Ready for review_ just to get the CI feedback, but it's _not_ ready for a review :-). Everything seems in-place though, just checking. The history needs to be rewritten.

File: crates/matrix-sdk/src/event_cache/mod.rs [hunk 1]
```diff
@@ -954,7 +1032,9 @@ mod tests {
             let room_id = room_id!("!galette:saucisse.bzh");
             client.base_client().get_or_create_room(room_id, matrix_sdk_base::RoomState::Joined);
 
-            client.event_cache().subscribe().unwrap();
+            let event_cache = client.event_cache();
+
+            event_cache.subscribe().unwrap();
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/mod.rs:1037]: Can you undo this change? The local isn't reused.

**Review by bnjbvr — CHANGES_REQUESTED** [review]
It's sweet to finally connect the linked chunks implementation to the event cache, nicely done!

A few comments below, and I'd like to take another look after those have been addressed. Thanks!

File: crates/matrix-sdk/src/event_cache/store.rs [hunk 2]
```diff
@@ -230,6 +50,11 @@ impl RoomEvents {
         Self { chunks: LinkedChunk::new() }
     }
 
+    /// Clear all events.
+    pub fn reset(&mut self) {
+        self.chunks = LinkedChunk::new();
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/store.rs:55]: Hmm I don't think it's worth the effort. Ultimately, the `reset` must not be used. We use it now because we don't the reconciliation algorithm yet, but it's not the final desired situation.

And recycling `Chunk` is going to be quite complex honestly… (since it's a pointer-based structure, the risk of messing things up is not zero 😛). I understand the reason, but let's not do some premature optimisation here.

File: crates/matrix-sdk/src/event_cache/store.rs [hunk 3]
```diff
@@ -251,12 +76,17 @@ impl RoomEvents {
         self.chunks.push_items_back(events)
     }
 
+    /// Push a gap after existing events.
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/store.rs:79]: I use the word _events_ in `RoomEvents` and _items` in `LinkedChunk`. In `LinkedChunk`, an item can be anything, but in `RoomEvents`, the items of `LinkedChunk` _are_ events.

File: crates/matrix-sdk/src/event_cache/store.rs [hunk 4]
```diff
@@ -265,14 +95,29 @@ impl RoomEvents {
...
         self.chunks.insert_gap_at(gap, position)
     }
 
+    /// Replace the gap identified by `gap_identifier`, by events.
+    ///
+    /// Because the `gap_identifier` can represent non-gap chunk, this method
+    /// returns a `Result`.
+    ///
+    /// The returned `Chunk` represents the newly created `Chunk` that contains
+    /// the first events.
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/store.rs:108]: Oh yeah, [that's the same comment as @poljar did already in another PR](https://github.com/matrix-org/matrix-rust-sdk/pull/3251#discussion_r1532051735). I forgot to update this comment too. Thanks for noticing. It's updated.

File: crates/matrix-sdk/tests/integration/event_cache.rs [hunk 5]
```diff
@@ -428,11 +428,12 @@ async fn test_reset_while_backpaginating() {
             JoinedRoomBuilder::new(room_id)
                 // Note to self: a timeline must have at least single event to be properly
                 // serialized.
-                .add_timeline_event(event_builder.make_sync_message_event(
+                .add_timeline_event(event_builder.make_sync_message_event_with_id(
                     user_id!("@a:b.c"),
+                    event_id!("$from_first_sync"),
                     RoomMessageEventContent::text_plain("heyo"),
                 ))
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/tests/integration/event_cache.rs]: Done

File: crates/matrix-sdk/tests/integration/event_cache.rs [hunk 6]
```diff
@@ -471,25 +472,27 @@ async fn test_reset_while_backpaginating() {
         JoinedRoomBuilder::new(room_id)
             // Note to self: a timeline must have at least single event to be properly
             // serialized.
-            .add_timeline_event(event_builder.make_sync_message_event(
+            .add_timeline_event(event_builder.make_sync_message_event_with_id(
                 user_id!("@a:b.c"),
+                event_id!("$from_second_sync"),
                 RoomMessageEventContent::text_plain("heyo"),
             ))
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/tests/integration/event_cache.rs]: Done

File: crates/matrix-sdk/tests/integration/event_cache.rs [hunk 7]
```diff
@@ -471,25 +472,27 @@ async fn test_reset_while_backpaginating() {
...
-            .set_timeline_prev_batch("second_backpagination".to_owned())
+            .set_timeline_prev_batch("second_prev_batch_token_from_sync".to_owned())
             .set_timeline_limited(),
     );
     let sync_response_body = sync_builder.build_json_sync_response();
 
     // First back-pagination request:
-    let chunk = non_sync_events!(event_builder, [ (room_id, "$2": "lalala") ]);
+    let chunk = non_sync_events!(event_builder, [ (room_id, "$from_backpagination": "lalala") ]);
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/tests/integration/event_cache.rs]: Done.

File: crates/matrix-sdk/tests/integration/event_cache.rs [hunk 8]
```diff
@@ -512,16 +515,17 @@ async fn test_reset_while_backpaginating() {
     mock_sync(&server, sync_response_body, None).await;
     client.sync_once(Default::default()).await.unwrap();
 
-    let outcome = backpagination.await.expect("join failed").unwrap();
+    let outcome = backpagination.await.expect("join failed");
 
     // Backpagination should be confused, and the operation should result in an
     // unknown token.
-    assert_matches!(outcome, BackPaginationOutcome::UnknownBackpaginationToken);
+    assert_matches!(outcome, Ok(BackPaginationOutcome::UnknownBackpaginationToken));
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/tests/integration/event_cache.rs]: Done.

File: crates/matrix-sdk/src/event_cache/mod.rs [hunk 9]
```diff
@@ -256,15 +249,14 @@ impl EventCache {
         // We could have received events during a previous sync; remove them all, since
         // we can't know where to insert the "initial events" with respect to
         // them.
-        let store = self.inner.store.lock().await;
+        // let mut room_events = room_cache.inner.events.write().await;
+        // room_events.reset();
 
-        store.clear_room(room_id).await?;
-        let _ = room_cache.inner.sender.send(RoomEventCacheUpdate::Clear);
+        // let _ = room_cache.inner.sender.send(RoomEventCacheUpdate::Clear);
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/mod.rs:255]: Thanks, done.

File: crates/matrix-sdk/src/event_cache/mod.rs [hunk 10]
```diff
@@ -282,20 +274,14 @@ struct EventCacheInner {
...
+    multiple_room_updates_lock: Mutex<()>,
+
     /// Lazily-filled cache of live [`RoomEventCache`], once per room.
     by_room: RwLock<BTreeMap<OwnedRoomId, RoomEventCache>>,
 
-    /// Backend used for storage.
-    ///
-    /// [`Mutex`] is “fair”, as it is implemented as a FIFO. It is important to
-    /// ensure that multiple updates will be applied in the correct order, which
-    /// is enforced by taking the store lock when handling an update.
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/mod.rs]: Indeed, done!

File: crates/matrix-sdk/src/event_cache/mod.rs [hunk 11]
```diff
@@ -205,14 +205,7 @@ impl EventCache {
...
                     // TODO: implement Smart Matching™,
-                    let store = inner.store.lock().await;
-                    let mut by_room = inner.by_room.write().await;
-                    for room_id in by_room.keys() {
-                        if let Err(err) = store.clear_room(room_id).await {
-                            error!("unable to clear room after room updates lag: {err}");
-                        }
-                    }
-                    by_room.clear();
+                    inner.by_room.write().await.clear();
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/mod.rs:208]: 1. Re-using chunks is not something I would like to do now, as explained in the other comment.
2. Why having a loop is better here? It's just slower for no particular benefits.

File: crates/matrix-sdk/src/event_cache/mod.rs [hunk 12]
```diff
@@ -505,51 +483,100 @@ impl RoomEventCacheInner {
...
-            let _ = self.sender.send(RoomEventCacheUpdate::Clear);
-        }
+    async fn handle_left_room_update(&self, updates: LeftRoomUpdate) -> Result<()> {
+        self.handle_timeline(updates.timeline, Vec::new(), Vec::new(), updates.ambiguity_changes)
+            .await?;
+        Ok(())
+    }
+
+    // Remove existing events, and append a set of events to the room cache and
+    // storage, notifying observers.
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/mod.rs:521]: What? There is a doc comment already.

File: crates/matrix-sdk/src/event_cache/mod.rs [hunk 13]
```diff
@@ -505,51 +483,100 @@ impl RoomEventCacheInner {
...
-        Ok(())
+        .await
     }
 
-    /// Append a set of events to the room cache and storage, notifying
-    /// observers.
-    async fn append_events(
+    /// Append a set of events, with an attached lock.
+    ///
+    /// If the lock `room_events` is `None`, one will be created.
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/mod.rs:574]: Yeah, it was my first design. But then I had doubt because if `events` is empty, `prev_batch` is empty`, `ephemeral`, `account_data` and `ambiguity_changes` are all empty, we simply return early. I don't know if it's a good idea to take the lock for no reason in this particular case.

We can also assume that this case is pretty rare as far as I know. I'm modifying the code according to your feedback, let's see how it goes.

File: crates/matrix-sdk/src/event_cache/mod.rs [hunk 14]
```diff
@@ -640,41 +684,68 @@ impl RoomEventCacheInner {
...
+
+        // There is a `token`/gap, let's replace it by new events!
+        if let Some(gap_identifier) = gap_identifier {
+            let new_position = {
+                // Replace the gap by new events.
+                let new_chunk = room_events
+                    .replace_gap_at(sync_events, gap_identifier)
+                    // SAFETY: we are sure that `gap_identifier` represents a valid
+                    // `ChunkIdentifier` for a gap.
+                    .unwrap();
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/mod.rs:698]: Yup Sir! Done.

File: crates/matrix-sdk/src/event_cache/mod.rs [hunk 15]
```diff
@@ -640,41 +684,68 @@ impl RoomEventCacheInner {
...
-        } else {
             trace!("replaced gap with new events from backpagination");
 
             // TODO: implement smarter reconciliation later
             //let _ = self.sender.send(RoomEventCacheUpdate::Prepend { events });
 
+            Ok(BackPaginationOutcome::Success { events, reached_start })
+        }
+        // There is no `token`/gap identifier. Let's assume we must prepend the new events.
+        else {
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/mod.rs:720]: It seems wrong to me, but OK.

File: crates/matrix-sdk/src/event_cache/mod.rs [hunk 16]
```diff
@@ -640,41 +684,68 @@ impl RoomEventCacheInner {
...
+            let first_item_position =
+                room_events.events().nth(0).map(|(item_position, _)| item_position);
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/mod.rs:722]: Indeed.

File: crates/matrix-sdk/src/event_cache/mod.rs [hunk 17]
```diff
@@ -689,22 +760,27 @@ impl RoomEventCacheInner {
...
+            })
+        }
+
+        if let Some(token) = get_oldest(self.events.read().await) {
             return Ok(Some(token));
         }
 
         let Some(max_wait) = max_wait else {
-            // We had no token and no time to wait, so... no tokens.
+            // We had no token and no time to wait, so… no tokens.
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/mod.rs:770]: 😅

File: crates/matrix-sdk/src/event_cache/mod.rs (hunk 1 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/mod.rs]: I forgot to change something else, it's done.

File: crates/matrix-sdk/src/event_cache/mod.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (bnjbvr) [reply crates/matrix-sdk/src/event_cache/mod.rs:521]: This is a non-doc comment:

```suggestion
    /// Remove existing events, and append a set of events to the room cache and
    /// storage, notifying observers.
```

File: crates/matrix-sdk/src/event_cache/mod.rs [hunk 18]
```diff
@@ -505,51 +484,99 @@ impl RoomEventCacheInner {
...
-        Ok(())
+        .await
     }
 
-    /// Append a set of events to the room cache and storage, notifying
-    /// observers.
-    async fn append_events(
+    /// Append a set of events, with an attached lock.
+    ///
+    /// If the lock `room_events` is `None`, one will be created.
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/mod.rs:574]: Thanks, I like the new API better :)

```suggestion
```

File: crates/matrix-sdk/src/event_cache/store.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (bnjbvr) [reply crates/matrix-sdk/src/event_cache/store.rs:79]: Theoretically, the last item of `self.chunks` could be a gap, so we could be pushing a gap after a gap? (even though we *should not*) In this case, `after existing events` would be misleading, because the pushed gap would go after a gap, not after an event.

**Review by bnjbvr — APPROVED** [review]
yay, great job, thank you!

File: crates/matrix-sdk/src/event_cache/store.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/store.rs:79]: It's possible to have multiple contiguous gap yes. I'll update the doc.

**Comment by Hywan** [comment]
I've rewritten the entire history of this patch. I think we are good!
